### ISSUE TITLE
High CPU load and choppy playback with 10 bit AVC encoded files (Kodi 21.1 Android and Android TV), but video is smooth in other players

### ISSUE THREAD

**Issue by sasha-spb-ru** [xbmc/xbmc#25870]
### Describe the bug
Here is a clear and concise description of what the problem is:

When playing some videos, there is a heavy CPU load and choppy playback (especially noticeable in scrolling and zooming scenes).
When watching on Android TV, the usual CPU load is about 30-50% (even at 4K), but when watching problematic videos, the load reaches 350% (even if the video is 720p).
When watching on my Android tablet, the usual CPU load is about 10-15% and 180-200% with problematic videos.
Сhanging various audio/video setting in Kodi does not result in any significant changes in CPU load when playing such files.
Such files are consistently observed in some releases (e.g. GameCenterCX videos, links to clippings from which I will give below).
But when playing same files on the same TV in other players, even system player (which can't be called a quality benchmark), the video plays perfectly smooth.
Test clips: [scrolling](https://drive.google.com/file/d/1bs8_XphETG6c7LVhOiplG8_unJBE884a/view?usp=sharing), [zooming](https://drive.google.com/file/d/17dX06OjHb64tLIFCx7EI47H6I-zOLmqA/view?usp=sharing)

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

The video should play smoothly as intended.
CPU utilization should be normal.

## Actual Behavior

Video playback is choppy.
CPU utilization is close to maximum across all cores, especially on weak devices and TVs.

### To Reproduce

1. Enable debug in settings and play the files from the links in description above using Kodi for Android (Android TV version is more preferable, as the CPU load is more noticeable there).
2. Monitor the CPU load.

### Debuglog

The debuglog can be found here:
Since this issue was created, the log has been updated to reflect the main viewing device - Sony X950H (LED TV flagship from 2020).
Full log with various audio/video component logging enabled - https://paste.kodi.tv/ufuliqumix.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [X] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Android 10 (Sony Android TV), Android 11 (tablet)
 - Kodi version: 21.1 Omega

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by sarbes** [comment]
I'd say your device has no hardware decoder for the codec of your video. And the CPU is too weak for the playback path we have available.

**Comment by sasha-spb-ru** [comment]
> I'd say your device has no hardware decoder for the codec of your video. And the CPU is too weak for the playback path we have available.

If my devices don't have this decoder, why does their system players and other player Apps play these video smoothly as it should be?

**Comment by sarbes** [comment]
Because they use a different playback path.

**Comment by sasha-spb-ru** [comment]
> Because they use a different playback path.

I admit I'm not an audio/video expert, but literally all the files I've looked at right now using MediaInfo tool are encoded with AVC (which is probably the standard now).
But so far only the examples I've cited have this problem with playback in Kodi.
Here is the exported info about one of the files:

General
Unique ID                       : 107021013563180602363361339564915741924 (0x50837F5A31A59D16604395E60524D4E4)
Complete name             : C:\Temp\Game Center CX 327 - Getsu Fuma Den [GooseCanyon].mkv
Format                           : Matroska
Format version              : Version 4
File size                          : 1.68 GiB
Duration                        : 57 min 29 s
Overall bit rate              : 4 176 kb/s
Frame rate                     : 59.940 FPS
Movie name                  : Game Center CX 327 - Getsu Fuma Den [GooseCanyon]
Encoded date               : 2022-11-24 09:51:12 UTC
Writing application      : mkvmerge v42.0.0 ('Overtime') 64-bit
Writing library              : libebml v1.3.10 + libmatroska v1.5.2 / Lavf58.76.100

Video
ID                                       : 1
Format                               : AVC
Format/Info                        : Advanced Video Codec
Format profile                    : High 10@L5
Format settings                  : CABAC / 16 Ref Frames
Format settings, CABAC     : Yes
Format settings, Reference frames        : 16 frames
Codec ID                            : V_MPEG4/ISO/AVC
Duration                             : 57 min 29 s
Bit rate                               : 3 919 kb/s
Width                                 : 1 280 pixels
Height                                : 720 pixels
Display aspect ratio            : 16:9
Frame rate mode                : Constant
Frame rate                           : 59.940 (59940/1000) FPS
Color space                         : YUV
Chroma subsampling          : 4:2:0
Bit depth                             : 10 bits
Scan type                            : Progressive
Bits/(Pixel*Frame)                : 0.071
Stream size                          : 1.57 GiB (94%)
Title                                     : HDTV - Skye
Writing library                      : x264 core 164 r3075 66a5bc1
Encoding settings                : cabac=1 / ref=16 / deblock=1:0:0 / analyse=0x3:0x113 / me=hex / subme=7 / psy=1 / psy_rd=1.00:0.00 / mixed_ref=1 / me_range=16 / chroma_me=1 / trellis=1 / 8x8dct=1 / cqm=0 / deadzone=21,11 / fast_pskip=1 / chroma_qp_offset=-2 / threads=12 / lookahead_threads=2 / sliced_threads=0 / nr=0 / decimate=1 / interlaced=0 / bluray_compat=0 / constrained_intra=0 / bframes=16 / b_pyramid=2 / b_adapt=1 / b_bias=0 / direct=1 / weightb=1 / open_gop=0 / weightp=2 / keyint=250 / keyint_min=25 / scenecut=40 / intra_refresh=0 / rc_lookahead=40 / rc=crf / mbtree=1 / crf=18.0 / qcomp=0.60 / qpmin=0 / qpmax=81 / qpstep=4 / ip_ratio=1.40 / aq=1:1.00
Language                              : Japanese
Default                                  : Yes
Forced                                   : No
Color range                           : Limited
Color primaries                      : BT.709
Transfer characteristics          : BT.709
Matrix coefficients                 : BT.709

Audio
ID                                       : 2
Format                               : AAC LC
Format/Info                        : Advanced Audio Codec Low Complexity
Codec ID                            : A_AAC-2
Duration                             : 57 min 29 s
Bit rate                                : 253 kb/s
Channel(s)                          : 2 channels
Channel layout                   : L R
Sampling rate                     : 48.0 kHz
Frame rate                           : 46.875 FPS (1024 SPF)
Compression mode             : Lossy
Delay relative to video        : 4 ms
Stream size                         : 104 MiB (6%)
Title                                    : .ts from anonymous
Language                            : Japanese
Default                                 : Yes
Forced                                  : No

Text
ID                                       : 3
Format                                : ASS
Codec ID                            : S_TEXT/ASS
Codec ID/Info                     : Advanced Sub Station Alpha
Duration                              : 57 min 27 s
Bit rate                                : 174 b/s
Frame rate                          : 0.350 FPS
Count of elements             : 1208
Compression mode           : Lossless
Stream size                         : 73.5 KiB (0%)
Default                                : Yes
Forced                                : No

**Comment by sarbes** [comment]
10 bit AVC is definitely not "the standard now". A lot of devices don't have hardware accelerators for it.

**Comment by sasha-spb-ru** [comment]
> A lot of devices don't have hardware accelerators for it.

I confirm that 10 bit AVC is the cause of the choppy playback and high CPU utilization.
But considering that for tablets and TV, random file manager from PlayStore and system video players play such codecs without problems, can we hope for some workaround in Kodi to play such files smoothly too?

**Comment by sasha-spb-ru** [comment]
User [Luke M](https://forum.kodi.tv/member.php?action=profile&uid=231521) from the Kodi forum gave me his [experimental build](https://drive.google.com/file/d/1Bh6B9qS5HHOlyacoY_GvnV83vS28Kbrj/view?usp=sharing).
And I confirm that with his version on my TV (Sony X950H) AVC High10 files are played smoothly and without CPU load.
So, the fix is definitely possible.
Here is a link to an [old thread (from 2021), description of the problem and solution](https://forum.kodi.tv/showthread.php?tid=336151&pid=3046728) (very simple).
I hope the fix can be implemented in the next official builds.
Full log from this build and playback of the previously problematic video file - https://paste.kodi.tv/ewonipiqah.kodi

**Comment by sarbes** [comment]
So it is not a Kodi bug. 

You might want to report it to the device manufacturer.

**Comment by joseluismarti** [comment]
@sasha-spb-ru There is a possible fix #25878, please try the [build](https://jenkins.kodi.tv/job/android-arm-docker/8320/artifact/kodi-20241025-1872e974-PR25878-armeabi-v7a.apk).

**Comment by sasha-spb-ru** [comment]
> There is a possible fix #25878, please try the [build](https://jenkins.kodi.tv/job/android-arm-docker/8320/artifact/kodi-20241025-1872e974-PR25878-armeabi-v7a.apk).

I tested this build on my TV - the new option works as intended.
Of course, on a device without support for this profile, enabling it causes Kodi to freeze, so option should be used with understanding.
Thank you very much for your work.
This issue has been closed, although I'm sure it's not only on these devices that this problem can be observed.
And it is almost impossible to bring the essence of the problem to such large manufacturers (tried many times).
It would be a pity if the PR is not accepted, but then I will have a reason to learn new skills for my own builds.
